functions/go/upsert-resource/upsertresource/upsert_resource.go

```diff
@@ -0,0 +1,167 @@
+		if IsSameResource(inputMeta, rMeta) {
+			nodes[i], err = deepCopy(inputResource)
+			if err != nil {
+				return false, err
+			}
+			a := combineInputAndMatchedAnnotations(inputMeta.Annotations, rMeta.Annotations)
+			err = nodes[i].SetAnnotations(a)
+			if err != nil {
+				return false, err
+			}
+			// found a matching resource
+			found = true
+		}
+	}
+	return found, nil
+}
+
+// AddResource appends the inputResource to the list of input nodes
+// it also cleans up the meta annotations so that resource is created in new file
+// by the function orchestrator
+func AddResource(nodes []*yaml.RNode, inputResource *yaml.RNode) ([]*yaml.RNode, error) {
+	new, err := deepCopy(inputResource)
+	if err != nil {
+		return nodes, err
+	}
+	meta, err := new.GetMeta()
+	if err != nil {
+		return nodes, err
+	}
+	// remove local, function, path and index annotations from the result
+	// removing path and index annotations makes orchestrator write resource
+	// to a new file
+	cleanedAnno := removeMetaAnnotations(meta.Annotations)
+	err = new.SetAnnotations(cleanedAnno)
+	if err != nil {
+		return nodes, err
+	}
+	nodes = append(nodes, new)
+	return nodes, nil
+}
+
+// IsSameResource returns true if metadata of two resources
+// have same Group, Kind, Name, Namespace
+func IsSameResource(meta1, meta2 yaml.ResourceMeta) bool {
+	g1, _ := ParseGroupVersion(meta1.APIVersion)
+	g2, _ := ParseGroupVersion(meta1.APIVersion)
+	return g1 == g2 && meta1.Kind == meta2.Kind &&
+		meta1.Name == meta2.Name &&
+		meta1.Namespace == meta2.Namespace
+}
+
+// ParseGroupVersion parses a KRM metadata apiVersion field.
+func ParseGroupVersion(apiVersion string) (group, version string) {
+	if i := strings.Index(apiVersion, "/"); i > -1 {
+		return apiVersion[:i], apiVersion[i+1:]
+	}
+	return "", apiVersion
+}
+
+// combineInputAndMatchedAnnotations combines user provided non-meta annotations from inputResource,
+// with path and index annotations from matched resource and returns the result
+func combineInputAndMatchedAnnotations(inputResourceAnno, matchedResourceAnno map[string]string) map[string]string {
+	if inputResourceAnno == nil {
+		inputResourceAnno = make(map[string]string)
+	}
+	if matchedResourceAnno == nil {
+		matchedResourceAnno = make(map[string]string)
+	}
+	res := make(map[string]string)
+	// retain the annotations from the input resource in fn-config,
+	// these should be written to matched resource
+	for k, v := range inputResourceAnno {
+		res[k] = v
+	}
+	// retain the path and index annotation from matched resource to result
+	res[kioutil.PathAnnotation] = matchedResourceAnno[kioutil.PathAnnotation]
+	res[kioutil.IndexAnnotation] = matchedResourceAnno[kioutil.IndexAnnotation]
+	// remove local and function meta annotations from the result as they should
+	// not be written to output resource
+	return removeLocalAndFnAnnotations(res)
+}
+
+// deepCopy returns the deep copy of the input RNode
+func deepCopy(node *yaml.RNode) (*yaml.RNode, error) {
```

Comment:
You are doing https://pkg.go.dev/sigs.k8s.io/kustomize/kyaml@v0.10.13/yaml#RNode.Copy.

Reply:
Thanks.